src/org/exist/storage/XQueryPool.java

```diff
@@ -58,6 +56,8 @@
 
     private final static Logger LOG = LogManager.getLogger(XQueryPool.class);
 
+    private ConcurrentMap<Source, Deque<CompiledXQuery>> pool = new ConcurrentHashMap<>();
+
     private long lastTimeOutCheck;
     private long lastTimeOfCleanup;
 
```

Comment:
Yes, however in the methods I described, your `time` is effectively `final` with regards to the other invariants so we need not consider it here.

`ConcurrentMap` doesn't have any impact on the `lastTimeOfCleanup` or `timeoutCheckInterval`, that has a different purpose. My point was that two threads where one pre-empts the other could now cause a different outcome on the values of `lastTimeOfCleanup` and `timeoutCheckInterval` because the two things are no longer synchronized together. That was all I was pointing out.

I don't see a problem with Deque, if you want to update it atomically, then just use the `compute...` functions on the `ConcurrentHashMap`.